assets/js/modules/tagmanager/util/validation.js

```diff
@@ -75,6 +75,57 @@ export function isValidContainerID( containerID ) {
 	return typeof containerID === 'string' && /^GTM-[A-Z0-9]+$/.test( containerID );
 }
 
+/**
+ * Gets normalized container name.
+ *
+ * @since n.e.x.t
+ *
+ * @param {string} containerName Container name to normalize.
+ * @return {string} Normalized container name.
+ */
+export function getNormalizedContainerName( containerName ) {
+	let sanitizedContainerName = containerName;
+
+	// Remove any leading or trailing whitespace.
+	sanitizedContainerName = sanitizedContainerName.trim();
+	// Must not start with an underscore.
+	sanitizedContainerName = sanitizedContainerName.replace( /^_+/, '' );
+	// Convert accents to basic characters to prevent them from being stripped.
+	sanitizedContainerName = sanitizedContainerName.normalize( 'NFD' ).replace( /[\u0300-\u036f]/g, '' );
+	// Strip all non-simple characters.
+	sanitizedContainerName = sanitizedContainerName.replace( /[^a-zA-Z0-9_., -]/g, '' );
+	// Collapse multiple whitespaces.
+	sanitizedContainerName = sanitizedContainerName.replace( /\s+/g, ' ' );
+
+	return sanitizedContainerName;
+}
+
+/**
+ * Checks if the given container name appears to be valid.
+ *
+ * @since n.e.x.t
+ *
+ * @param {string} containerName Container name to test.
+ * @return {boolean} True if valid, otherwise false.
+ */
+export function isValidContainerName( containerName ) {
+	return typeof containerName === 'string' && getNormalizedContainerName( containerName ).length > 0;
+}
+
+/**
+ * Checks if the given cotnainer name is unique accross account containers.
+ *
+ * @since n.e.x.t
+ *
+ * @param {string}         containerName Container name to test.
+ * @param {Array.<Object>} containers    Available containers.
+ * @return {boolean} True if unique, otherwise false.
+ */
+export function isUniqueContainerName( containerName, containers ) {
+	const normalizedContainerName = getNormalizedContainerName( containerName );
+	return ! Array.isArray( containers ) || ! containers.some( ( { name } ) => name === normalizedContainerName );
```

Comment:
Would you please add a few tests for this?

I think the logic is slightly off here
```suggestion
	return Array.isArray( containers ) && ! containers.some( ( { name } ) => name === normalizedContainerName );
```

Reply:
The logic is fine, it just accounts for a case when containers aren't loaded yet and `containers` value is undefined.

Reply:
That makes more sense, thanks for clarifying.

Not sure if you considered this before, but I think this might be better implemented as a private selector in `containers.js`. Then we could simply have `select(...).isUniqueContainerName( name )` and it would select all containers internally and return `undefined` if containers were not loaded yet for the current account. That way it would be less prone to improper use and more accurate in regards to the behavior in `canSubmitChanges` (an `undefined` return value from the strict selector would throw). What do you think?

Reply:
I like how it is implemented now. The `isUniqueContainerName` function is intended for validation, not for selecting something from the store. So, I don't see it in the datastore itself.

Reply:
It's not _bad_ the way it is now, my suggestion was that I think it may make more sense as a selector than a utility function. The main reason why I think that is because it depends on all the containers in datastore for the current account. Not only that, but it wouldn't make sense to use otherwise – there's no other case where we would need to check if a container name was unique among a different set of containers. So rather than putting the burden on the consumer to get the `accountID`, and select all containers properly to provide to the utility, that can all be done in a single selector.

Then the usage becomes much more simple:
```diff
diff --git a/assets/js/modules/tagmanager/components/common/ContainerNameTextField.js b/assets/js/modules/tagmanager/components/common/ContainerNameTextField.js
index ce5155479..206efb937 100644
--- a/assets/js/modules/tagmanager/components/common/ContainerNameTextField.js
+++ b/assets/js/modules/tagmanager/components/common/ContainerNameTextField.js
@@ -34,22 +34,18 @@ import Data from 'googlesitekit-data';
 import { STORE_NAME, FORM_SETUP } from '../../datastore/constants';
 import { STORE_NAME as CORE_FORMS } from '../../../../googlesitekit/datastore/forms/constants';
 import { TextField, HelperText, Input } from '../../../../material-components';
-import { isUniqueContainerName } from '../../util';
 const { useSelect, useDispatch } = Data;
 
 export default function ContainerNameTextField( { label, name } ) {
-	const containers = useSelect( ( select ) => {
-		const accountID = select( STORE_NAME ).getAccountID();
-		return select( STORE_NAME ).getContainers( accountID );
-	} );
 	const containerName = useSelect( ( select ) => select( CORE_FORMS ).getValue( FORM_SETUP, name ) );
+	const isUniqueContainerName = useSelect( ( select ) => select( STORE_NAME ).isUniqueContainerName( containerName ) );
 
 	const { setValues } = useDispatch( CORE_FORMS );
 	const onChange = useCallback( ( { currentTarget } ) => {
 		setValues( FORM_SETUP, { [ name ]: currentTarget.value } );
 	}, [ name ] );
 
-	const helperText = containerName && ! isUniqueContainerName( containerName, containers )
+	const helperText = containerName && ! isUniqueContainerName
 		? <HelperText persistent>{ __( 'A container with this name already exists.', 'google-site-kit' ) }</HelperText>
 		: undefined;
```

The utility function itself could still exist as it is, the selector by the same name would simply use it internally.

Reply:
@aaemnnosttv I understand what you suggest and my opinion remains the same: I don't like the idea of putting validator functions into a datastore because it is out of scope for datastores. Putting all together breaks the "single responsibility" principle.

Following your logic, we would also need to move `isValidAccountID`, `isValidContainerID`, `isValidInternalContainerID`, and other validation functions into datastore as well because they all rely on data from a datastore and it would be easier to select just `isValidAccountID` selector than `getAccountID() + isValidAccountID()`.

I think consistency is more important than usability in this case.

Reply:
> Following your logic, we would also need to move `isValidAccountID`, `isValidContainerID`, `isValidInternalContainerID`, and other validation functions into datastore as well because they all rely on data from a datastore and it would be easier to select just `isValidAccountID` selector than selecting an account ID with `getAccountID()` and calling `isValidAccountID()` function.

No, that's not the same. We use those to validate provided arguments/props in many places where the source of the data is unknown – they aren't only used to validate a value which is already in the datastore.

This isn't a validation function, it's a selector 😄  I'm not suggesting we add `isValidContainerName` as a selector on the store, that would definitely be out of place. This would only be asking the store if it has a container with the given name; that's it. Since this value is only relevant in the context of all containers for the currently selected account, I think it makes sense as a selector, again particularly because it isn't useful with different/unknown values for `containers`; we specifically care about all of the containers for the current account.

> I think consistency is more important than usability in this case.

I don't think it's a tradeoff between the two. In a way, it becomes more consistent as a selector than only as a utility, but again, the two aren't mutually exclusive, we can have both 😄

Reply:
@eugene-manuilov – After thinking about this a bit more, this shouldn't block the PR from being merged, but I do think it is an enhancement we should apply in a follow-up issue. Another issue is needed anyways regarding the edge case I mentioned previously for secondary AMP, so we can continue the discussion in the definition of that issue.

Reply:
Sounds good @aaemnnosttv